File: src/main/java/work/xeltica/craft/core/utils/Time.java
package work.xeltica.craft.core.utils;

public class Time {
    public static String msToString(long timeInMilliseconds) {
        final var _ms = timeInMilliseconds;
        final var ms = _ms % 1000;
        final var _s = _ms / 1000;
        final var s = _s % 60;
        final var _m = _s / 60;
        final var m = _m % 60;
        final var _h = _m / 60;
        final var h = _h % 60;
        return h == 0
            ? String.format("%02d:%02d.%03d", m, s, ms)
            : String.format("%02d:%02d:%02d.%03d", h, m, s, ms);
    }
}

File: src/main/java/work/xeltica/craft/core/api/Time.kt
package work.xeltica.craft.core.api

object Time {
    @JvmStatic
    fun msToString(timeInMilliseconds: Long): String {
        val ms = timeInMilliseconds % 1000
        val _s = timeInMilliseconds / 1000
        val s = _s % 60
        val _m = _s / 60
        val m = _m % 60
        val _h = _m / 60
        val h = _h % 60
        return if (h == 0L) String.format(
            "%02d:%02d.%03d",
            m,
            s,
            ms
        ) else String.format("%02d:%02d:%02d.%03d", h, m, s, ms)
    }
}